File: WQChartsDemo/Objc/View/SectionCell.m
// 代码地址: 
// SectionCell.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "SectionCell.h"

@interface SectionCell()

@property (nonatomic, strong) UILabel* textLabel;
@property (nonatomic, strong) UIButton* button;

@end

@implementation SectionCell

#pragma mark - Public

- (instancetype)initWithFrame:(CGRect)frame {
    if(self = [super initWithFrame:frame]) {
        self.backgroundColor = UIColor.clearColor;
        [self addSubview:self.textLabel];
        [self addSubview:self.button];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    CGFloat width = self.bounds.size.width;
    CGFloat height = self.bounds.size.height;
    
    UIButton* button = self.button;
    CGSize buttonSize = CGSizeMake(40, 25);
    button.frame = CGRectMake(width - buttonSize.width - 10, (height - buttonSize.height) / 2, buttonSize.width, buttonSize.height);
    
    UILabel* textLabel = self.textLabel;
    [textLabel sizeToFit];
    CGSize textSize = textLabel.bounds.size;
    textLabel.frame = CGRectMake(10, (height - textSize.height) / 2, CGRectGetMinX(button.frame) - 20, textSize.height);
}

- (CGSize)sizeThatFits:(CGSize)size {
    return CGSizeMake(size.width, MIN(size.height, 40));
}

#pragma mark - Getter

- (UILabel *)textLabel {
    if(!_textLabel) {
        _textLabel = [[UILabel alloc] init];
        _textLabel.font = [UIFont systemFontOfSize:9];
        _textLabel.textColor = Color_White;
    }
    return _textLabel;
}

- (UIButton *)button {
    if(!_button) {
        _button = [UIButton buttonWithType:UIButtonTypeCustom];
        _button.backgroundColor = Color_White;
        _button.titleLabel.font = [UIFont systemFontOfSize:9];
        [_button setTitleColor:Color_Tint forState:UIControlStateNormal];
        [_button setTitle:@"Reload" forState:UIControlStateNormal];
        [_button addTarget:self action:@selector(onReloadButtonClick) forControlEvents:UIControlEventTouchUpInside];
    }
    return _button;
}

- (NSString *)title {
    return _textLabel.text;
}

#pragma mark - Setter

- (void)setTitle:(NSString *)title {
    self.textLabel.text = title;
    [self setNeedsLayout];
}

- (SectionCell*(^)(id))setObject {
    return ^SectionCell*(id object) {
        self.objcect = object;
        return self;
    };
}

- (SectionCell*(^)(NSString*))setTitle {
    return ^SectionCell*(NSString* title) {
        self.title = title;
        return self;
    };
}

- (SectionCell*(^)(SectionCellReloadHandler))setOnReload {
    return ^SectionCell*(SectionCellReloadHandler onReload) {
        self.onReload = onReload;
        return self;
    };
}

- (void)onReloadButtonClick {
    if(self.onReload) {
        self.onReload(self);
    }
}

@end

File: WQChartsDemo/Swift/View/SectionCell.swift
// 代码地址: https://github.com/CoderWQYao/WQCharts-iOS
//
// SectionCell.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class SectionCell: UIView {
    
    // MARK: - Property
    
    open lazy var textLabel: UILabel = {
        let textLabel = UILabel()
        textLabel.font = UIFont.systemFont(ofSize: 9)
        textLabel.textColor = Color_White
        return textLabel
    }()
    
    open lazy var button: UIButton = {
        let button = UIButton(type: .custom)
        button.backgroundColor = Color_White
        button.titleLabel?.font = UIFont.systemFont(ofSize: 9)
        button.setTitleColor(Color_Tint, for: .normal)
        button.setTitle("Reload", for: .normal)
        button.addTarget(self, action: #selector(callButtonClick), for: .touchUpInside)
        return button
    }()
    
    
    open var object: Any?

    open func setObject(_ object: Any?) -> SectionCell {
        self.object = object
        return self
    }
    
    open var title: String? {
        get {
            return textLabel.text
        } set {
            textLabel.text = newValue
            setNeedsLayout()
        }
    }
    
    open func setTitle(_ title: String?) -> SectionCell {
        textLabel.text = title
        return self
    }
    
    open var onReload: ((_ cell: SectionCell) -> Void)?
    
    open func setOnReload(_ onReload: ((_ cell: SectionCell) -> Void)?) -> SectionCell {
       self.onReload = onReload
       return self
    }
    
    // MARK: - Public
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        backgroundColor = .clear
        addSubview(textLabel)
        addSubview(button)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        let width = bounds.width
        let height = bounds.height
        
        let button = self.button
        let buttonSize = CGSize(width: 40, height: 25)
        button.frame = CGRect(x: width - buttonSize.width - 10, y: (height - buttonSize.height) / 2, width: buttonSize.width, height: buttonSize.height)
        
        let textLabel = self.textLabel
        textLabel.sizeToFit()
        let textSize = textLabel.bounds.size
        textLabel.frame = CGRect(x: 10, y: (height - textSize.height) / 2, width: button.frame.minX - 20, height: textSize.height)
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        return CGSize(width: size.width, height: min(size.height,40))
    }
    
    @objc open func callButtonClick() {
        if let onReload = onReload {
            onReload(self)
        }
    }
    
    
}
